### ISSUE TITLE
[DataGrid] onFilterModelChange is not fired when last character in value field is deleted

### ISSUE THREAD

**Issue by h-tomek** [mui-org/material-ui-x#2940]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Current behavior 😯

If we have a single character written in filter value input box, and then deleted it, the `filterModelChange` event is not fired, thus, the `onFilterModelChange` callback is not executed.

This has been tested on latest Chrome, Firefox and Edge browsers, and the behaviour is the same on all of them.

### Expected behavior 🤔

One would expect that the event is being fired since there actually is a change in the filter model.

### Steps to reproduce 🕹

Steps:

1. Open filter
2. Enter a character in the input value field
3. filterModelChange event is fired
4. Delete the character entered in the step 2
5. filterModelChange is not fired

### Your environment 🌎

<details>
  <summary>`npx @mui/envinfo`</summary>

```
   System:
    OS: Windows 10 10.0.19042
    CPU: (12) x64 Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
    Memory: 1.19 GB / 15.74 GB
  Binaries:
    Node: 14.15.1 - C:\Program Files\nodejs\node.EXE
    npm: 6.14.8 - C:\Program Files\nodejs\npm.CMD
  Utilities:
    Git: 2.20.1. - /mingw64/bin/git
  Languages:
    Bash: 4.4.23 - C:\Program Files\Git\usr\bin\bash.EXE
    Java: 1.8.0_201 - /c/Program Files/Java/jdk1.8.0_201/bin/javac
    Perl: 5.26.2 - C:\Program Files\Git\usr\bin\perl.EXE
  Databases:
    PostgreSQL: 10.6 - C:\PostgreSQL\pg10\bin\postgres.EXE
  Browsers:
    Chrome: 94.0.4606.81
    Edge: Spartan (44.19041.1266.0), Chromium (94.0.992.50)
    Internet Explorer: 11.0.19041.1202
```
</details>

### Order ID 💳 (optional)

#23200

**Comment by oliviertassinari** [comment]
@DanailH How did you try to reproduce?

I have tried "Steps to reproduce 🕹" provided on the latest version: v5.0.0-beta.6 and on HEAD, it's indeed broken. It's not only about `onFilterModelChange` that doesn't fire, but the filter that doesn't reset itself. End-users are stuck, they can't disable the filter by emptying the input.
https://codesandbox.io/s/serverfiltergrid-material-demo-forked-ueyzq?file=/demo.tsx:0-1888

So I had a quick look at what happened. It's a regression from v5.0.0-beta.3. It's a regression from #2780. cc @m4theushw.

**Comment by m4theushw** [comment]
This will be fixed by #3018
